File: benchmark.erl
-module(benchmark).
-export([main/1]).

% 内部ループの計算を行う関数
calc_inner_sum(U) ->
    % 0からU-1までの余りの合計を計算
    BaseSum = lists:sum(lists:seq(0, U-1)),
    
    % 完全なU個のグループの数と余りを計算
    Quotient = 99999 div U,
    Remainder = 99999 rem U,
    
    % 合計を計算
    RemainderSum = lists:sum([X rem U || X <- lists:seq(0, Remainder)]),
    
    BaseSum * Quotient + RemainderSum.

% 配列の一部を計算するワーカープロセス
worker(Parent, InnerSum, R, Start, End) ->
    Result = array:from_list([InnerSum + R || _ <- lists:seq(Start, End-1)]),
    Parent ! {self(), Result}.

main([Arg]) ->
    try
        U = list_to_integer(Arg),
        
        % 乱数生成
        <<A:32, B:32, C:32>> = crypto:strong_rand_bytes(12),
        rand:seed(exs1024s, {A,B,C}),
        R = rand:uniform(10000) - 1,
        
        % 内部ループの計算を1回だけ行う
        InnerSum = calc_inner_sum(U),
        
        % プロセッサ数に基づいてワーカー数を決定
        WorkerCount = erlang:system_info(schedulers_online),
        ChunkSize = (10000 + WorkerCount - 1) div WorkerCount,
        
        % ワーカープロセスを起動
        Workers = lists:map(
            fun(N) ->
                Start = N * ChunkSize,
                End = min(Start + ChunkSize, 10000),
                Pid = spawn_link(fun() -> worker(self(), InnerSum, R, Start, End) end),
                {Pid, Start, End}
            end,
            lists:seq(0, WorkerCount-1)
        ),
        
        % 結果を収集
        Array = lists:foldl(
            fun({Pid, Start, End}, Acc) ->
                receive
                    {Pid, Result} ->
                        array:set_range(Start, array:to_list(Result), Acc)
                end
            end,
            array:new(10000),
            Workers
        ),
        
        % 結果を出力
        io:format("~p~n", [array:get(R, Array)]),
        erlang:halt(0)
    catch
        _:_ ->
            io:format(standard_error, "Usage: escript benchmark.erl <number>~n", []),
            erlang:halt(1)
    end;
main(_) ->
    io:format(standard_error, "Usage: escript benchmark.erl <number>~n", []),
    erlang:halt(1).

File: benchmark.rs
use std::env;
use rand::Rng;

// 内部合計を計算する関数
fn calc_inner_sum(u: i32) -> i32 {
    // 0からu-1までの余りの合計を計算
    let base_sum: i32 = (0..u).sum();
    
    // 完全なu個のグループの数と余りを計算
    let quotient = 99999 / u;
    let remainder = 99999 % u;
    
    // 合計を計算（Rustのイテレータを活用）
    let total = base_sum * quotient +
                (0..=remainder).map(|x| x % u).sum::<i32>();
    
    total
}

fn main() {
    // コマンドライン引数の処理
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <number>", args[0]);
        std::process::exit(1);
    }

    let u: i32 = args[1].parse().expect("Invalid number");
    
    // 乱数生成
    let mut rng = rand::thread_rng();
    let r: usize = rng.gen_range(0..10000);
    
    // 内部ループの計算を1回だけ行う
    let inner_sum = calc_inner_sum(u);
    
    // ベクタの初期化を最適化（Rustのメモリ安全性を活用）
    let a: Vec<i32> = vec![inner_sum + r as i32; 10000];
    
    println!("{}", a[r]);
}
